### ISSUE TITLE
[ENH] - Add Explainability Module for sktime Models

### ISSUE THREAD

**Issue by Thomas4390** [sktime/sktime#6949]
**Is your feature request related to a problem? Please describe.**

sktime currently lacks built-in tools for model explainability, making it difficult for users to interpret and understand the predictions generated by time series models. This is a crucial feature, especially in domains where trust and transparency are essential for decision-making.

**Describe the solution you'd like**

I propose the addition of an explainability module within sktime that would allow users to easily interpret the outputs of their models. This could include features like SHAP values, permutation importance, and other interpretability tools, similar to those implemented in the skforecast library. The solution should integrate seamlessly with sktime’s existing classes and methods.

**Describe alternatives you've considered**

While external libraries like SHAP or LIME could be used, they require additional adaptation for time series models, which can be complex and less user-friendly. Integrating these capabilities directly into sktime would offer a more streamlined and cohesive user experience.

**Additional context**

Implementing this feature would greatly enhance the utility of sktime, providing users with essential tools to better understand and trust their model predictions.

**Comment by fkiraly** [comment]
Interesting idea!

I assume you mean general purpose interpretability tools, where you can analyze the fitted estimator and its predictions, for, e.g., "why was this prediction made" type of analysis?

This seems like a major feature request, but also sth that has repeatedly come up: see, e.g., https://github.com/sktime/sktime/issues/4336

What would be very useful:

* pointers to general purpose algorithms in the time series domain
* pointers to algorithms that are specific to individual algorithms - e.g., shap can be naively applied to reduction algorithm such as those provided to `skforecast` or `sktime` or `darts`. Algorithm specific interpretability modes can be added to the individual algorithm, there is a tag for "has feature importance" - https://www.sktime.net/en/stable/api_reference/auto_generated/sktime.registry._tags.capability__feature_importance.html

What might be good first steps:

* shap etc can be easily (but not trivially) generalized to time series classifiers, regressors, etc. This could be added along the lines in #4336
* similarly, this could be done for algorithms that reduce to tabular

More advanced question are, how would the API design look like, as we are dealing with a variety of different estimator types, e.g., for forecasting? Inspiration could be taken from libraries such as `fairlearn`.

Any other thoughts?

**Comment by Thomas4390** [comment]
First of all, thank you for considering this idea and for your continued work on sktime!

It is exactly as you said: "general purpose interpretability tools" or also "model explainability" to analyze the fitted estimator and its predictions. I think you've already got an excellent idea of the different issues that this kind of module can entail for sktime. 

Speaking only for SHAP analysis, the two main ingredients are:

- The internal regressor of the forecaster (after fit).
- The training matrices created from the time series and used to fit the forecaster.

See [skforecast_model_explanaibility](https://skforecast.org/0.13.0/user_guides/explainability#shap-explainer) for a more complete explanation. 

Integrating these interpretability tools could greatly enhance sktime's utility, making it an even more valuable resource for practitioners. I’m excited to see where this discussion leads!

**Comment by fkiraly** [comment]
> The internal regressor of the forecaster (after fit).

That's what I mean with "reduction" - this approach works only if there is such an internal regressor, and that basically means we are looking at reduction models (direct or recursive), i.e., the kind of model that `skforecast` implements.

Most forecasting models are *not* reduciton models, even if reduction models are a common and not unpopular choice.

I.e., it is relatively straightforward what to do for reduction models, but not clear what one should to for arbitrary forecasting models.

**Comment by Thomas4390** [comment]
I agree with you! This seems to be more of a case-by-case approach for forecasting models without reduction (e.g. with ARIMA).

**Comment by fkiraly** [comment]
> I agree with you! This seems to be more of a case-by-case approach for forecasting models without reduction (e.g. with ARIMA).

Not necessarily? There are general purpose methods that one could think of, e.g., simple input perturbation strategies?

**Comment by Thomas4390** [comment]
Hello @fkiraly, I hope this message finds you well. I just wanted to follow up and ask if there have been any updates on this matter since our last discussion. Thank you again for all your contributions to the open-source community.
